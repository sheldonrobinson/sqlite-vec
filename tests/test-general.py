import sqlite3
from collections import OrderedDict
import pytest


@pytest.mark.skipif(
    sqlite3.sqlite_version_info[1] < 37,
    reason="pragma_table_list was added in SQLite 3.37",
)
def test_shadow(db, snapshot):
    db.execute(
        "create virtual table v using vec0(a float[1], partition text partition key, metadata text, +name text, chunk_size=8)"
    )
    assert exec(db, "select * from sqlite_master order by name") == snapshot()
    assert (
        exec(db, "select * from pragma_table_list where type = 'shadow'") == snapshot()
    )

    db.execute("drop table v;")
    assert (
        exec(db, "select * from pragma_table_list where type = 'shadow'") == snapshot()
    )


def test_info(db, snapshot):
    db.execute("create virtual table v using vec0(a float[1])")
    assert exec(db, "select key, typeof(value) from v_info order by 1") == snapshot()


def exec(db, sql, parameters=[]):
    try:
        rows = db.execute(sql, parameters).fetchall()
    except (sqlite3.OperationalError, sqlite3.DatabaseError) as e:
        return {
            "error": e.__class__.__name__,
            "message": str(e),
        }
    a = []
    for row in rows:
        o = OrderedDict()
        for k in row.keys():
            o[k] = row[k]
        a.append(o)
    result = OrderedDict()
    result["sql"] = sql
    result["rows"] = a
    return result


def vec0_shadow_table_contents(db, v):
    shadow_tables = [
        row[0]
        for row in db.execute(
            "select name from sqlite_master where name like ? order by 1", [f"{v}_%"]
        ).fetchall()
    ]
    o = {}
    for shadow_table in shadow_tables:
        o[shadow_table] = exec(db, f"select * from {shadow_table}")
    return o
