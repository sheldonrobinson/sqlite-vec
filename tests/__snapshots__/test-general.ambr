# serializer version: 1
# name: test_info
  OrderedDict({
    'sql': 'select key, typeof(value) from v_info order by 1',
    'rows': list([
      OrderedDict({
        'key': 'CREATE_VERSION',
        'typeof(value)': 'text',
      }),
      OrderedDict({
        'key': 'CREATE_VERSION_MAJOR',
        'typeof(value)': 'integer',
      }),
      OrderedDict({
        'key': 'CREATE_VERSION_MINOR',
        'typeof(value)': 'integer',
      }),
      OrderedDict({
        'key': 'CREATE_VERSION_PATCH',
        'typeof(value)': 'integer',
      }),
    ]),
  })
# ---
# name: test_shadow
  OrderedDict({
    'sql': 'select * from sqlite_master order by name',
    'rows': list([
      OrderedDict({
        'type': 'index',
        'name': 'sqlite_autoindex_v_info_1',
        'tbl_name': 'v_info',
        'rootpage': 3,
        'sql': None,
      }),
      OrderedDict({
        'type': 'index',
        'name': 'sqlite_autoindex_v_metadatachunks00_1',
        'tbl_name': 'v_metadatachunks00',
        'rootpage': 10,
        'sql': None,
      }),
      OrderedDict({
        'type': 'index',
        'name': 'sqlite_autoindex_v_metadatatext00_1',
        'tbl_name': 'v_metadatatext00',
        'rootpage': 12,
        'sql': None,
      }),
      OrderedDict({
        'type': 'index',
        'name': 'sqlite_autoindex_v_vector_chunks00_1',
        'tbl_name': 'v_vector_chunks00',
        'rootpage': 8,
        'sql': None,
      }),
      OrderedDict({
        'type': 'table',
        'name': 'sqlite_sequence',
        'tbl_name': 'sqlite_sequence',
        'rootpage': 5,
        'sql': 'CREATE TABLE sqlite_sequence(name,seq)',
      }),
      OrderedDict({
        'type': 'table',
        'name': 'v',
        'tbl_name': 'v',
        'rootpage': 0,
        'sql': 'CREATE VIRTUAL TABLE v using vec0(a float[1], partition text partition key, metadata text, +name text, chunk_size=8)',
      }),
      OrderedDict({
        'type': 'table',
        'name': 'v_auxiliary',
        'tbl_name': 'v_auxiliary',
        'rootpage': 13,
        'sql': 'CREATE TABLE "v_auxiliary"( rowid integer PRIMARY KEY , value00)',
      }),
      OrderedDict({
        'type': 'table',
        'name': 'v_chunks',
        'tbl_name': 'v_chunks',
        'rootpage': 4,
        'sql': 'CREATE TABLE "v_chunks"(chunk_id INTEGER PRIMARY KEY AUTOINCREMENT,size INTEGER NOT NULL,sequence_id integer,partition00,validity BLOB NOT NULL, rowids BLOB NOT NULL)',
      }),
      OrderedDict({
        'type': 'table',
        'name': 'v_info',
        'tbl_name': 'v_info',
        'rootpage': 2,
        'sql': 'CREATE TABLE "v_info" (key text primary key, value any)',
      }),
      OrderedDict({
        'type': 'table',
        'name': 'v_metadatachunks00',
        'tbl_name': 'v_metadatachunks00',
        'rootpage': 9,
        'sql': 'CREATE TABLE "v_metadatachunks00"(rowid PRIMARY KEY, data BLOB NOT NULL)',
      }),
      OrderedDict({
        'type': 'table',
        'name': 'v_metadatatext00',
        'tbl_name': 'v_metadatatext00',
        'rootpage': 11,
        'sql': 'CREATE TABLE "v_metadatatext00"(rowid PRIMARY KEY, data TEXT)',
      }),
      OrderedDict({
        'type': 'table',
        'name': 'v_rowids',
        'tbl_name': 'v_rowids',
        'rootpage': 6,
        'sql': 'CREATE TABLE "v_rowids"(rowid INTEGER PRIMARY KEY AUTOINCREMENT,id,chunk_id INTEGER,chunk_offset INTEGER)',
      }),
      OrderedDict({
        'type': 'table',
        'name': 'v_vector_chunks00',
        'tbl_name': 'v_vector_chunks00',
        'rootpage': 7,
        'sql': 'CREATE TABLE "v_vector_chunks00"(rowid PRIMARY KEY,vectors BLOB NOT NULL)',
      }),
    ]),
  })
# ---
# name: test_shadow.1
  OrderedDict({
    'sql': "select * from pragma_table_list where type = 'shadow'",
    'rows': list([
      OrderedDict({
        'schema': 'main',
        'name': 'v_auxiliary',
        'type': 'shadow',
        'ncol': 2,
        'wr': 0,
        'strict': 0,
      }),
      OrderedDict({
        'schema': 'main',
        'name': 'v_chunks',
        'type': 'shadow',
        'ncol': 6,
        'wr': 0,
        'strict': 0,
      }),
      OrderedDict({
        'schema': 'main',
        'name': 'v_info',
        'type': 'shadow',
        'ncol': 2,
        'wr': 0,
        'strict': 0,
      }),
      OrderedDict({
        'schema': 'main',
        'name': 'v_rowids',
        'type': 'shadow',
        'ncol': 4,
        'wr': 0,
        'strict': 0,
      }),
      OrderedDict({
        'schema': 'main',
        'name': 'v_metadatachunks00',
        'type': 'shadow',
        'ncol': 2,
        'wr': 0,
        'strict': 0,
      }),
      OrderedDict({
        'schema': 'main',
        'name': 'v_metadatatext00',
        'type': 'shadow',
        'ncol': 2,
        'wr': 0,
        'strict': 0,
      }),
    ]),
  })
# ---
# name: test_shadow.2
  OrderedDict({
    'sql': "select * from pragma_table_list where type = 'shadow'",
    'rows': list([
    ]),
  })
# ---
